README.md
# React Idle Timer

![npm](https://img.shields.io/npm/v/react-idle-timer.svg)
![Travis](https://img.shields.io/travis/SupremeTechnopriest/react-idle-timer.svg)
![Code Climate](https://img.shields.io/codeclimate/coverage/SupremeTechnopriest/react-idle-timer.svg)
![Code Climate](https://img.shields.io/codeclimate/maintainability/SupremeTechnopriest/react-idle-timer.svg)
![Code Climate](https://img.shields.io/codeclimate/tech-debt/SupremeTechnopriest/react-idle-timer.svg)

[![NPM](https://nodei.co/npm/react-idle-timer.png?downloads=true&stars=true)](https://npmjs.org/package/react-idle-timer/)
[![JavaScript Style Guide](https://cdn.rawgit.com/standard/standard/master/badge.svg)](https://github.com/standard/standard)

⚡️ **Support for React 16**<br/>
🚀 **Support for Isomorphic React**

## Latest News

Welcome to version 4 of IdleTimer! We have performed a complete rewrite of our build system and a refactor/ clean up of the source code. After accepting many pull requests things started to get ugly. We added test coverage and continuous integration tools (travis and codeclimate) that will automatically enforce style and test future pull requests.  

There is one breaking change in version 4:

- The property `startOnLoad` has been renamed to `startOnMount` in order to make more sense in a react context.

For the full patch notes please refer to the [CHANGELOG](https://github.com/SupremeTechnopriest/react-idle-timer/blob/master/example/src/CHANGELOG.md)

## Installation
`yarn add react-idle-timer`
or
`npm install react-idle-timer --save`

## Usage

> Run `yarn example` to build and run the example `example`. The example is a [create-react-app](https://github.com/facebook/create-react-app) project. IdleTimer is implemented in the [App Component](https://github.com/SupremeTechnopriest/react-idle-timer/blob/master/example/src/App.js).

```javascript
import React, { Component } from 'react'
import IdleTimer from 'react-idle-timer'

export default class YourApp extends Component {
  constructor(props) {
    super(props)
    this.idleTimer = null
    this.onActive = this._onActive.bind(this)
    this.onIdle = this._onIdle.bind(this)
  }

  render() {
    return (
      <IdleTimer
        ref={ref => { this.idleTimer = ref }}
        element={document}
        activeAction={this.onActive}
        idleAction={this.onIdle}
        timeout={1000 * 60 * 15}>

        <h1>Child Components</h1>

      </IdleTimer>
    )
  }

  _onActive(e) {
    console.log('user is active', e)
    console.log('time remaining', this.idleTimer.getRemainingTime())
  }

  _onIdle(e) {
    console.log('user is idle', e)
    console.log('last active', this.idleTimer.getLastActiveTime())
  }
}
```

## Documentation

> To build the source code generated html docs run `yarn docs` and open `docs/index.html` in any browser.  A markdown version is available [here](https://github.com/SupremeTechnopriest/react-idle-timer/blob/master/example/src/DOCS.md).

### Default Events
These events are bound by default:
- mousemove
- keydown
- wheel
- DOMMouseScroll
- mouseWheel
- mousedown
- touchstart
- touchmove
- MSPointerDown
- MSPointerMove

### Props
- **timeout** {*Number*} - Idle timeout in milliseconds
- **events** {*Array*} - Events to bind. See [default events](https://)
- **idleAction** {*Function*} - Function to call on idle
- **activeAction** {*Function*} - Function to call on active
- **element** {*Object*} - Defaults to document, may pass a ref to another element
- **startOnMount** {*Boolean*} - Start the timer on component load.  Defaults to `true`. Set to false to wait for user action before starting timer.
- **passive** {*Boolean*} - Bind events in passive mode. Defaults to `true`. Set to false to explicitly set to active mode.
- **capture** {*Boolean*} - Capture events. Defaults to `true`

### Methods
- **reset()** *{Void}* - Resets the idleTimer
- **pause()** *{Void}* - Pauses the idleTimer
- **resume()** *{Void}* - Resumes a paused idleTimer
- **getRemainingTime()** *{Number}* - Returns the remaining time in milliseconds
- **getElapsedTime()** *{Number}* - Returns the elapsed time in milliseconds
- **getLastActiveTime()** *{String}* - Returns the `Date` the user was last active
- **isIdle()** *{Boolean}* - Returns whether or not user is idle


CHANGES.md
## Breaking Changes
### Props
startOnLoad -> startOnMount


src/index.js
/**
 *  ___    _ _     _____ _
 * |_ _|__| | | __|_   _(_)_ __ ___   ___ _ __
 *  | |/ _` | |/ _ \| | | | '_ ` _ \ / _ \ '__|
 *  | | (_| | |  __/| | | | | | | | |  __/ |
 * |___\__,_|_|\___||_| |_|_| |_| |_|\___|_|
 *
 * @name IdleTimer
 * @author Randy Lebeau
 * @private
 */

import { Component } from 'react'
import PropTypes from 'prop-types'

/**
 * Determine if we are in a browser
 * or a server environment
 * @type {Boolean}
 * @private
 */
const IS_BROWSER = (typeof window === 'undefined' ? 'undefined' : typeof (window)) === 'object'

/**
 * Default element to listen for events on
 * @type {Object}
 * @private
 */
const DEFAULT_ELEMENT = IS_BROWSER ? document : {}

/**
 * The default events to determine activity
 * @type {Array}
 * @private
 */
const DEFAULT_EVENTS = [
  'mousemove',
  'keydown',
  'wheel',
  'DOMMouseScroll',
  'mouseWheel',
  'mousedown',
  'touchstart',
  'touchmove',
  'MSPointerDown',
  'MSPointerMove'
]

/**
 * Detects when your user is idle
 * @class IdleTimer
 * @private
 */
export default class IdleTimer extends Component {
  /**
   * Type checks for every property
   * @type {Object}
   * @private
   */
  static propTypes = {
    /**
     * Activity Timeout
     * @type {Number}
     */
    timeout: PropTypes.number,
    /**
     * Activty events to listen on
     * @type {Array}
     */
    events: PropTypes.arrayOf(PropTypes.string),
    /**
     * Function to call when user is idle
     * @type {Function}
     */
    idleAction: PropTypes.func,
    /**
     * Function to call when user becomes active
     * @type {Function}
     */
    activeAction: PropTypes.func,
    /**
     * Element reference to bind activity listeners to
     * @type {Object}
     */
    element: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
    /**
     * Start the timer on mount
     * @type {Boolean}
     */
    startOnMount: PropTypes.bool,
    /**
     * Bind events passively
     * @type {Boolean}
     */
    passive: PropTypes.bool,
    /**
     * Capture events
     * @type {Boolean}
     */
    capture: PropTypes.bool
  }

  /**
   * Sets default property values
   * @type {Object}
   * @private
   */
  static defaultProps = {
    timeout: 1000 * 60 * 20,
    element: DEFAULT_ELEMENT,
    events: DEFAULT_EVENTS,
    idleAction: () => {},
    activeAction: () => {},
    startOnMount: true,
    capture: true,
    passive: true
  }

  /**
   * Sets initial component state
   * @type {Object}
   * @private
   */
  state = {
    idle: false,
    oldDate: +new Date(),
    lastActive: +new Date(),
    remaining: null,
    pageX: null,
    pageY: null
  }

  /**
   * The timer instance
   * @type {Timeout}
   * @private
   */
  tId = null

  /**
   * Creates an instance of IdleTimer
   * bind all of our internal events here
   * for best performance
   * @param {Object} props
   * @return {IdleTimer}
   * @private
   */
  constructor (props) {
    super(props)
    this.toggleIdleState = this._toggleIdleState.bind(this)
    this.reset = this._reset.bind(this)
    this.pause = this._pause.bind(this)
    this.resume = this._resume.bind(this)
    this.getRemainingTime = this._getRemainingTime.bind(this)
    this.getElapsedTime = this._getElapsedTime.bind(this)
    this.getLastActiveTime = this._getLastActiveTime.bind(this)
    this.isIdle = this._isIdle.bind(this)
  }

  /**
   * Runs when the component mounts
   * here we bind the events to the
   * element we want to listen on
   * @private
   */
  componentWillMount () {
    // Dont bind events if
    // we are not in a browser
    if (!IS_BROWSER) return
    // Otherwise we bind all the events
    // to the supplied element
    const { element, events, passive, capture } = this.props
    events.forEach(e => element.addEventListener(e, this._handleEvent), {
      capture,
      passive
    })
  }

  /**
   * Runs once the component has mounted
   * here we handle automatically starting
   * the idletimer
   * @private
   */
  componentDidMount () {
    // If startOnMount is enabled
    // start the timer
    const { startOnMount } = this.props
    if (startOnMount) {
      this.reset()
    }
  }

  /**
   * Called before the component unmounts
   * here we clear the timer and remove
   * all the event listeners
   * @private
   */
  componentWillUnmount () {
    // Clear timeout to prevent delayed state changes
    clearTimeout(this.tId)
    // If we are not in a browser
    // we dont need to unbind events
    if (!IS_BROWSER) return
    // Unbind all events
    const { element, events } = this.props
    events.forEach(e => element.removeEventListener(e, this._handleEvent))
  }

  /**
   * Render children if IdleTimer is used as a wrapper
   * @return {Component} children
   * @private
   */
  render () {
    const { children } = this.props
    return children || null
  }

  /**
   * Toggles the idle state and calls
   * the correct action function
   * @private
   */
  _toggleIdleState (e) {
    // Toggle the idle state
    const { idle } = this.state
    this.setState({
      idle: !idle
    })

    // Fire the appropriate action
    // and pass the event through
    const { activeAction, idleAction } = this.props
    if (idle) {
      activeAction(e)
    } else {
      idleAction(e)
    }
  }

  /**
   * Event handler for supported event types
   * @param  {Object} e event object
   * @private
   */
  _handleEvent = (e) => {
    const { remaining, pageX, pageY } = this.state
    // Already idle, ignore events
    if (remaining) return

    // Mousemove event
    if (e.type === 'mousemove') {
      // If coord are same, it didn't move
      if (e.pageX === pageX && e.pageY === pageY) {
        return
      }
      // If coord don't exist how could it move
      if (typeof e.pageX === 'undefined' && typeof e.pageY === 'undefined') {
        return
      }
      // Under 200 ms is hard to do
      // continuous activity will bypass this
      const elapsed = this.getElapsedTime()
      if (elapsed < 200) {
        return
      }
    }

    // Clear any existing timeout
    clearTimeout(this.tId)

    // If the idle timer is enabled, flip
    if (this.state.idle) {
      this._toggleIdleState(e)
    }

    // Store when the user was last active
    // and update the mouse coordinates
    this.setState({
      lastActive: +new Date(), // store when user was last active
      pageX: e.pageX, // update mouse coord
      pageY: e.pageY
    })

    // Set a new timeout
    const { timeout } = this.props
    this.tId = setTimeout(this._toggleIdleState.bind(this), timeout) // set a new timeout
  }

  /**
   * Restore initial state and restart timer
   * @name reset
   */
  _reset () {
    // Clear timeout
    clearTimeout(this.tId)

    // Reset state
    this.setState({
      idle: false,
      oldDate: +new Date(),
      lastActive: this.state.oldDate,
      remaining: null
    })

    // Set new timeout
    const { timeout } = this.props
    this.tId = setTimeout(this._toggleIdleState.bind(this), timeout)
  }

  /**
   * Store remaining time and stop timer
   * @name pause
   */
  _pause () {
    // Timer is already paused
    const { remaining } = this.state
    if (remaining !== null) {
      return
    }

    // Clear existing timeout
    clearTimeout(this.tId)

    // Define how much is left on the timer
    this.setState({
      remaining: this.getRemainingTime()
    })
  }

  /**
   * Resumes a paused timer
   * @name resume
   */
  _resume () {
    // Timer is not paused
    const { remaining, idle } = this.state
    if (remaining === null) {
      return
    }

    // Start timer and clear remaining
    // if we are in the idle state
    if (!idle) {
      this.setState({ remaining: null })
      // Set a new timeout
      this.tId = setTimeout(this._toggleIdleState.bind(this), remaining)
    }
  }

  /**
   * Time remaining before idle
   * @name getRemainingTime
   * @return {Number} Milliseconds remaining
   */
  _getRemainingTime () {
    const { remaining, idle, lastActive } = this.state
    // If idle there is no time remaining
    if (idle) {
      return 0
    }

    // If timer is in a paused state
    // just return its remaining time
    if (remaining !== null) {
      return remaining
    }

    // Determine remaining, if negative idle didn't finish flipping, just return 0
    const { timeout } = this.props
    let timeLeft = timeout - ((+new Date()) - lastActive)
    if (timeLeft < 0) {
      timeLeft = 0
    }
    return timeLeft
  }

  /**
   * How much time has elapsed
   * @name getElapsedTime
   * @return {Timestamp}
   */
  _getElapsedTime () {
    const { oldDate } = this.state
    return (+new Date()) - oldDate
  }

  /**
   * Last time the user was active
   * @name getLastActiveTime
   * @return {Timestamp}
   */
  _getLastActiveTime () {
    const { lastActive } = this.state
    return lastActive
  }

  /**
   * Returns wether or not the user is idle
   * @name isIdle
   * @return {Boolean}
   */
  _isIdle () {
    const { idle } = this.state
    return idle
  }
}
